presto-cluster-ttl-providers/src/main/java/com/facebook/presto/clusterttlproviders/percentile/PercentileBasedClusterTtlProvider.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.clusterttlproviders.percentile;
+
+import com.facebook.presto.spi.ttl.ClusterTtlProvider;
+import com.facebook.presto.spi.ttl.ConfidenceBasedTtlInfo;
+import com.facebook.presto.spi.ttl.NodeTtl;
+import com.google.inject.Inject;
+
+import java.util.Comparator;
+import java.util.List;
+import java.util.Optional;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.toList;
+
+public class PercentileBasedClusterTtlProvider
+        implements ClusterTtlProvider
+{
+    private final Comparator<ConfidenceBasedTtlInfo> ttlComparator = Comparator.comparing(ConfidenceBasedTtlInfo::getExpiryInstant);
+    private final int percentile;
+
+    @Inject
+    public PercentileBasedClusterTtlProvider(PercentileBasedClusterTtlProviderConfig config)
+    {
+        requireNonNull(config, "config is null");
+        checkArgument(config.getPercentile() > 0, "percentile should be greater than 0");
+        this.percentile = config.getPercentile();
+    }
+
+    @Override
+    public ConfidenceBasedTtlInfo getClusterTtl(List<NodeTtl> nodeTtls)
+    {
+        List<ConfidenceBasedTtlInfo> ttls = nodeTtls.stream()
+                .map(this::getMinTtl)
+                .filter(Optional::isPresent)
+                .map(Optional::get)
+                .sorted(ttlComparator)
+                .collect(toList());
+
+        if (ttls.size() == 0) {
+            return new ConfidenceBasedTtlInfo(0, 100);
+        }
+
+        return ttls.get((int) Math.ceil(ttls.size() * percentile / 100.0) - 1);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The result of `Integer` multiplication is cast into a larger type. If you think the result of the multiplication could exceed the size of an `Integer`, consider casting the arguments of the multiplication to avoid arithmetic overflows. For more information, see [CWE-190](https://cwe.mitre.org/data/definitions/190.html).